### PR TITLE
add gvisor based service library 

### PR DESCRIPTION
**Pull request by d4l3k**
This adds a gvisor based service library which allows for using a user space TCP stack instead of a native tun device.

This is an initial implementation and only handles TCP and ipv4 connections. This implementation is a bit rough so feedback is welcome :slightly_smiling_face:

This refactors the Main function to take in a DeviceFactory -- so any users that require tunFd need to use NewFdDeviceFromConfig(tunFd) as the factory instead

See examples/go_service/main.go for how this might be used

Test plan:

```
$ go test ./service
```

Manual
```
$ go run ./examples/go_service
$ socat 192.168.100.2 -
hello world
asdf
echo: "asdf"
duck
echo: "duck"
```

### CODE REVIEW

**Review by brad-defined — COMMENTED** [review]
only a couple minor comments from what I saw. I think we can land this PR and update it as needed with future PRs.

Other thoughts (which I don't think should be required for the PR to land)
- It'd be fun to write some performance tests
- I expect the io.Pipe used doesn't scale up, but this can be updated in a future PR (along with performance tests that can verify or disprove my suspicion.)

File: service/service.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	"gvisor.dev/gvisor/pkg/tcpip/transport/icmp"
+	"gvisor.dev/gvisor/pkg/tcpip/transport/tcp"
+	"gvisor.dev/gvisor/pkg/tcpip/transport/udp"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+const nicID = 1
+
+type Service struct {
+	ctx     context.Context
```
[the remark this answers is not in the document]
Comment (d4l3k) [reply service/service.go]: done

File: service/service.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	}); err != nil {
+		return nil, fmt.Errorf("error creating IP: %s", err)
+	}
+
+	const tcpReceiveBufferSize = 0
+	const maxInFlightConnectionAttempts = 1024
+	tcpFwd := tcp.NewForwarder(s.ipstack, tcpReceiveBufferSize, maxInFlightConnectionAttempts, s.tcpHandler)
+	s.ipstack.SetTransportProtocolHandler(tcp.ProtocolNumber, tcpFwd.HandlePacket)
+
+	reader, writer := device.Pipe()
```
[the remark this answers is not in the document]
Comment (d4l3k) [reply service/service.go:109]: done

**Comment by brad-defined** [comment]
I'm OK with this landing. It adds some nice, useful functionality, modifies the Main() interface a little, and doesn't appear to me to add any interfaces that we can't live with going forward.

Will see if @wadey or @nbrownus have any thoughts on it.

**Comment by brad-defined** [comment]
>  Error: service/service.go:111:10: undefined: context.AfterFunc

My bad...apparently context.AfterFunc in go 1.21.0, but Nebula isn't on that golang version yet; it reduces Windows supported O/S's. Sorry! Will need to implement that functionality without that context function...

File: service/service.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+	context.AfterFunc(ctx, func() { reader.Close(); writer.Close() })
```
Comment (brad-defined) [inline service/service.go:111]:
```suggestion
	go func() {
		<-ctx.Done()
		reader.Close()
		writer.Close()
	}()
```

**Comment by johnmaguire** [comment]
I believe this PR would close #538.

**Review by brad-defined — APPROVED** [review]

**Comment by d4l3k** [comment]
@maggie44 I'm currently using this for my own (limited) use and it's been working just fine. You're right that ipv6 isn't supported by Nebula currently so that's just behavior as usual so just UDP isn't supported.

I've been using this for SSH and HTTP connections without any issues

**Pull request merged**